presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSource.java

```diff
@@ -164,18 +178,20 @@ public void finish()
         checkNotHoldsLock();
 
         SettableFuture<?> notEmptyFuture;
-        synchronized (lock) {
-            if (finishing) {
+        synchronized (this) {
+            if (!finishing.compareAndSet(false, true)) {
                 return;
             }
-            finishing = true;
 
+            //  Unblock any waiters
             notEmptyFuture = this.notEmptyFuture;
-            this.notEmptyFuture = NOT_EMPTY;
+            this.notEmptyFuture = null;
         }
 
         // notify readers outside of lock since this may result in a callback
-        notEmptyFuture.set(null);
+        if (notEmptyFuture != null) {
+            notEmptyFuture.set(null);
+        }
```

Comment:
Any reason making notEmptyFuture nullable is better than what we had before? This seems bulkier without notable benefit?

Reply:
Compared to what was happening before, this is indeed better in terms of performance: 
- Overwriting `this.notEmptyFuture` unconditionally is an additional (potentially irrelevant) store operation that must be flushed to memory as part of releasing the monitor. Skipping the store if the field is already null (ie: no listener is registered) avoids that.
- Unconditionally calling `SettableFuture#set(null)` in the previous implementation required a volatile CAS even when no listener was registered which performs poorly compared to a null check (which are exceedingly cheap)

I think the previous use of having the `static final SettableFuture<?> NOT_EMPTY` constant was arguable _less_ clear about what was going on. Conceptually, this field has two states: listener registered and no listener registered. You could make them _appear_ the same by using the constant, but then other parts of code are always checking `isDone()` which is expensive (volatile read) and in my opinion creates a hazard for race conditions since `set(null)` can race with the `isDone()` checks without paying close attention to subtle details of what happens inside of the syncronized sections.